src/utils/commands.ts

```diff
@@ -0,0 +1,71 @@
+import { Message, Permissions } from 'discord.js';
+import Commando, { CommandoMessage, CommandoClient, CommandInfo, ArgumentCollectorResult } from 'discord.js-commando';
+import logger, { logError } from '../components/logger';
+
+export abstract class BaseCommand extends Commando.Command {
+  constructor(client: CommandoClient, info: CommandInfo) {
+    super(client, info);
+  }
+
+  run(
+    message: CommandoMessage,
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    args: object | string | string[],
+    fromPattern: boolean,
+    result?: ArgumentCollectorResult
+  ): Promise<Message | Message[] | null> | null {
+    // log each incoming command
+    onCommandRun(message);
+    return this.onRun(message, args, fromPattern, result);
+  }
+
+  onError(
+    err: Error,
+    message: CommandoMessage,
+    _args: unknown,
+    _fromPattern: boolean,
+    _result?: ArgumentCollectorResult
+  ): Promise<Message> {
+    return handleCommandError(err, message);
+  }
+
+  abstract onRun(
+    message: CommandoMessage,
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    args: object | string | string[],
+    fromPattern: boolean,
+    result?: ArgumentCollectorResult
+  ): Promise<Message | Message[] | null> | null;
```

Comment:
How does a return value of null (without the promise) actually do anything? I was under the impression that the wrapped promise<null> is sufficient.

Reply:
When you have an async function, the return type has to be wrapped in a promise. `null` is just for non-async `onRun` where you return nothing.

Reply:
Now I think about it, I'm gonna remove `null` anyways, since Commando's underlying implementation is async so we should do the same